### PR TITLE
Better Item Selection UI

### PR DESCRIPTION
**Pull request by MEEPofFaith**
Edit: Here, have a demonstration video conveniently placed at the top so that you don't need to scroll down through the commits and comments.

https://user-images.githubusercontent.com/54301439/132141942-d2b55bc1-0b50-4637-97e0-78ebee0f9579.mp4

---

Due to the sheer amount of things you can select (especially with mods installed), I decided to make the Payload Source's selection UI wider.
Wide UI

If your pull request is **not** translation or serverlist-related, read the list of requirements below and check each box:

- [X] I have read the [contribution guidelines](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md).
- [X] I have ensured that my code compiles, if applicable.
- [X] I have ensured that any new features in this PR function correctly in-game, if applicable.

### CODE REVIEW

**Comment by MEEPofFaith** [comment]
Smaller UIs not affected.

https://user-images.githubusercontent.com/54301439/131821040-d50de1ac-a194-4d7e-871f-0d9b58ec9412.mp4

**Comment by MEEPofFaith** [comment]
I found that each row was 53 1/3 tall, and used that in the row calculations.

**Comment by sk7725** [comment]
test it in non-100% ui scaling

**Comment by USMP-lancer** [comment]
> test it in non-100% ui scaling

yes

**Comment by MEEPofFaith** [comment]
_Good point, I play on 75% UI scale, dunno about other scales_

**Comment by MEEPofFaith** [comment]
aaa this is uncomfortable
image

**Comment by MEEPofFaith** [comment]
image
uhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

Yea I'll make "8 rows" be based off of 100%

**Comment by MEEPofFaith** [comment]
Ok, made the rows based on 100% UI scaling so that it's correct.

https://user-images.githubusercontent.com/54301439/131893957-2f742995-eeb0-473a-ae6d-e0e6be65f5ad.mp4

~~_I am uncomfortable with how large everything is._~~

**Comment by USMP-lancer** [comment]
> ~_I am uncomfortable with how large everything is._~
~reverse compensation~

**Comment by MEEPofFaith** [comment]
This footage is in 100% UI scale, I've tested and it works fine in other UI scales.

https://user-images.githubusercontent.com/54301439/131934778-4a5ee007-040e-4e91-ac6c-3ee1d037de38.mp4


Mods active to bloat the amount of possible selections:
- BetaMindy
- Exotic Mod
- Exogenesis
- Endless Rusting
- Progressed Materials

**Comment by MEEPofFaith** [comment]
image
Now to fix it in other UI scales...

**Comment by MEEPofFaith** [comment]
Why didn't this exist before?
https://user-images.githubusercontent.com/54301439/131936453-599f878c-6e56-4a49-8d13-bb01bbc45b1c.mp4

(Also I'm back down to a comfortable 75% UI scale)

**Comment by MEEPofFaith** [comment]
Made the displayed images larger so that their easier to see and easier to actually highlight over the image to make the name tooltip appear.

https://user-images.githubusercontent.com/54301439/131937114-c11e2e53-9fd6-420d-94f9-c386982a5caa.mp4

**Comment by MEEPofFaith** [comment]
> image
> Now to fix it in other UI scales...
> 
> (This is 25%)

The width of the text box doesn't change with the scaling, and I don't know how to make it do that.

**Comment by buthed010203** [comment]
I tried to make it work with scaling and gave up, im convinced its impossible.
What you can try tho is setting the width to Scl.scl(40 * columns), it works ok but isnt perfect

**Comment by MEEPofFaith** [comment]
image

**Comment by buthed010203** [comment]
Why is rebuild a Runnable[] of length 1 rather than just Runnable rebuild = ...? both approaches work the same

**Comment by MEEPofFaith** [comment]
> Why is rebuild a Runnable[] of length 1 rather than just Runnable rebuild = ...? both approaches work the same

unsure, but the attribute stats on AttributeCrafter use a single term array as well

**Comment by MEEPofFaith** [comment]
Now making the ui smaller doesn't change the number of rows visible at once
image

**Comment by MEEPofFaith** [comment]
25% is so tiny why does it exist

**Comment by MEEPofFaith** [comment]
I think that's everything done :D

https://user-images.githubusercontent.com/54301439/132075852-96b013cb-9c7c-49b6-897d-9629b1984a54.mp4

**Comment by USMP-lancer** [comment]
~~commit more changes i dare you~~

**Comment by MEEPofFaith** [comment]
~~There's nothing left to commit~~

**Comment by buthed010203** [comment]
> I think that's everything done :D
> 
>  Final.mp4

Ohno, search bar still too large

**Comment by USMP-lancer** [comment]
> > I think that's everything done :D
> > Final.mp4
> 
> Ohno, search bar still too large

amazing

**Comment by MEEPofFaith** [comment]
> > I think that's everything done :D
> > Final.mp4
> 
> Ohno, search bar still too large

wdym search bar too large

> Now changing the ui scale doesn't change the number of rows visible at once
> image

**Comment by sk7725** [comment]
> I think that's everything done :D
> 
>  Final.mp4

uh it should snap to the top, not to the center

**Comment by MEEPofFaith** [comment]
> > I think that's everything done :D
> > Final.mp4
> 
> uh it should snap to the top, not to the center

even with `.top()` it does not snap to the top, I've already tried

**Comment by MEEPofFaith** [comment]
image
image
help

**Comment by MEEPofFaith** [comment]
There, snapping to top.

https://user-images.githubusercontent.com/54301439/132113032-e072387d-8c82-4080-8b85-c1014c465911.mp4

I wish I could make it not have the issue where scrolling down on a larger selection would make it severely overscroll on a smaller one, but `maxY` isn't set yet at that part, so trying to use `setScrollPercentY` would always scroll to the top because `maxY` is 0 when it's called.

**Comment by MEEPofFaith** [comment]
image

**Comment by MEEPofFaith** [comment]
*The background comes with an issue and I don't know how to fix it.*

image

**Comment by MEEPofFaith** [comment]
Solution: Who needs the search icon?
image
image

**Comment by buthed010203** [comment]
The line is not centered in the background 😠

**Comment by MEEPofFaith** [comment]
> The line is not centered in the background 😠

Cry about it
get real

**Comment by buthed010203** [comment]
> > The line is not centered in the background 😠
> 
> Cry about it
> get real

Ok i will

**Review by USMP-lancer — APPROVED** [review]
i understand none of it so i approve lmao

**Comment by MEEPofFaith** [comment]
> i understand none of it so i approve lmao

~~There is no need to understand, only to view my demo video~

**Comment by USMP-lancer** [comment]
instructions not clear, timetraveled and cause a plane hijacking in september 2004

**Comment by SoMall-dumpling** [comment]
LGTM.

**Comment by buthed010203** [comment]
Very cool

**Review by ChetanAtrawalkarCA — COMMENTED** [review]
Mean while the UI is improve and game is nicely build.

**Comment by MEEPofFaith** [comment]
~~Merge when~~

**Comment by buthed010203** [comment]
No

**Comment by MEEPofFaith** [comment]
> No
> ~might want to ping anuke considering he gave feedback thats since been implemented~

Well I did post in pulls and Anuke has talked about it.
image

**Comment by buthed010203** [comment]
Ah ok

**Review by buthed010203 — APPROVED** [review]

**Comment by MEEPofFaith** [comment]
Anything missing that I should add to this?

**Comment by sk7725** [comment]
> Anything missing that I should add to this? ~I'm too lazy to scroll down and look for stuff~

Maybe set the threshold to 1.5x the scrollpane (ceil it), not just the x1 the scrollpane height;
e.g. for item selection pane (height = 5), the search bat would show up if you have more than 8 rows of items (5 * 1.5 ceil = 8).

**Comment by MEEPofFaith** [comment]
makes sense, you don't need to scroll a lot

File: core/src/mindustry/world/Block.java [hunk 1]
```diff
@@ -167,6 +167,8 @@
     public boolean configurable;
     /** If true, the building inventory can be shown with the config. */
     public boolean allowConfigInventory = true;
+    /** Used in blocks with selection menus such sorters for how large the menu should be */
```
Comment (buthed010203) [inline core/src/mindustry/world/Block.java:170]: Such sorters

**Comment by MEEPofFaith** [comment]
~~I swear to god if this has private branch conflicts.~~

**Comment by buthed010203** [comment]
~~I will end up merging them when v7 comes out either way~~

**Comment by EclipseTheOldOne** [comment]
this is a 6 month post wha-

**Comment by MEEPofFaith** [comment]
~~I am surprised there are no merge conflicts~~

**Comment by buthed010203** [comment]
> I am surprised there are no merge conflicts

~~just wait for the private branch~~

**Comment by Anuken** [comment]
I could try merging this soon, I haven't modified the selection UI at all.

**Comment by Anuken** [comment]
Merged with a bunch of tweaks.

**Comment by MEEPofFaith** [comment]
epic

**Pull request closed without merging**